head/tests/test_presence_sensing_driver.cpp

```diff
@@ -18,4 +20,81 @@ SCENARIO("get readings called on presence sensing driver") {
+            auto raw_readings = adc::RawADCReadings{
+                .z_motor = 2332, .a_motor = 3548, .gripper = 666};
+            static auto adc_comms = adc::MockADC{raw_readings};
+            auto ps = presence_sensing_driver::PresenceSensingDriver(
+                adc_comms, ot3_tool_list::AttachedTool{});
+            auto tools = ot3_tool_list::AttachedTool(
+                ps.get_readings(), ot3_tool_list::get_tool_list());
+
+            THEN("Tools mapped to voltage reading") {
+                REQUIRE(tools.gripper == can_ids::ToolType::gripper);
+                REQUIRE(tools.a_motor == can_ids::ToolType::pipette_multi_chan);
+                REQUIRE(tools.z_motor == can_ids::ToolType::pipette_384_chan);
+            }
+        }
+        WHEN("get_tool func is called and raw ADC readings are out of bounds") {
+            auto raw_readings = adc::RawADCReadings{
+                .z_motor = 9999, .a_motor = 9999, .gripper = 9999};
+            static auto adc_comms = adc::MockADC{raw_readings};
+            auto ps = presence_sensing_driver::PresenceSensingDriver(
+                adc_comms, ot3_tool_list::AttachedTool{});
+            auto tools = ot3_tool_list::AttachedTool(
+                ps.get_readings(), ot3_tool_list::get_tool_list());
+
+            THEN("Tools mapped to voltage reading") {
+                REQUIRE(tools.gripper == can_ids::ToolType::undefined_tool);
+                REQUIRE(tools.a_motor == can_ids::ToolType::undefined_tool);
+                REQUIRE(tools.z_motor == can_ids::ToolType::undefined_tool);
+            }
+        }
+        WHEN(
+            "get_tool func is called and raw ADC readings are for no tool "
+            "attached") {
+            auto raw_readings =
+                adc::RawADCReadings{.z_motor = 2, .a_motor = 20, .gripper = 20};
+            static auto adc_comms = adc::MockADC{raw_readings};
+            auto ps = presence_sensing_driver::PresenceSensingDriver(
+                adc_comms, ot3_tool_list::AttachedTool{});
+            auto tools = ot3_tool_list::AttachedTool(
+                ps.get_readings(), ot3_tool_list::get_tool_list());
+
+            THEN("Tools mapped to voltage reading") {
+                REQUIRE(tools.gripper == can_ids::ToolType::nothing_attached);
+                REQUIRE(tools.a_motor == can_ids::ToolType::nothing_attached);
+                REQUIRE(tools.z_motor == can_ids::ToolType::nothing_attached);
+            }
+        }
+    }
+}
+
+SCENARIO("ot3 tools list for tool detection") {
+    GIVEN("ot3 tools lower and upper voltage bounds for tool detection") {
+        auto tools = ot3_tool_list::get_tool_list();
+        WHEN("when lower bound compared with upper bound") {
+            THEN("lower bound is always found lower or equal to upper bounds") {
+                for (auto& element : tools) {
+                    REQUIRE(element.bounds.lower <= element.bounds.upper);
```

Comment:
nice, good idea!

Reply:
was @amitlissack's idea. Thanks @amitlissack

Reply:
Hmmm. That wasn't really my idea. I was suggesting a concept that would require a Tool to have lower be less than upper.

This works too! SO it was all your idea, @sotaoverride !